echo-webhooks/src/main/groovy/com/netflix/spinnaker/echo/controllers/WebhooksController.groovy

```diff
@@ -140,31 +138,43 @@ class WebhooksController {
                                     @RequestBody CloudEvent cdevent,
                                     @RequestHeader HttpHeaders headers) {
     log.info("CDEvents Webhook received with source ${source} and with event type ${cdevent.getType()}")
-
-    String ceDataJsonString = new String(cdevent.getData().toBytes(), StandardCharsets.UTF_8);
-
-    Event event = new Event()
-    boolean sendEvent = true
-    event.details = new Metadata()
-    event.details.source = source
-    event.details.type = "cdevents"
-    event.details.requestHeaders = headers
-    event.rawContent = ceDataJsonString
+    String ceDataJsonString  = headers.get("Ce-Data").get(0)
+    log.info("CDEvent received with in the CloudEvent data request {}", ceDataJsonString)
     Map postedEvent
     try {
       postedEvent = mapper.readValue(ceDataJsonString, Map) ?: [:]
+      if (postedEvent.get("customData") == null || !postedEvent.get("context") || !postedEvent.get("subject")) {
```

Comment:
We're missing a number of tests around this logic here it seems like.  Be good to have some better tests on this, and maybe a doc stating why this is "invalid CDEvent" data?  ALSO... "RuntimeException" would be an incorrect exception given spinnaker standards.

Reply:
There is a validation for `artifacts` and `parameters` in the existing [test](https://github.com/spinnaker/echo/pull/1357/files#diff-3d61ce6cb6a561b259741a1af9c4616d5f34f6250551c94b56f0bfe57eaaacffR888-R889) with valid CDEvents data(will separate each case with different test, If needed ?) and there is a new test to verify invalid CDEvents data.

In the place of "RuntimeException" I can change to `com.netflix.spinnaker.kork.web.exceptions.InvalidRequestException` but I need to add a dependency of `io.spinnaker.kork:kork-web` in the echo-webhooks.gradle project, If that is fine?

I have create a [PR](https://github.com/spinnaker/spinnaker.io/pull/392), user guide to trigger the pipeline using CDEvents, where I mentioned how valid CDEvents data can be  constructed as per [CDEvents-spec](https://github.com/cdevents/spec/blob/main/spec.md)

Reply:
The existing test does a lot in ONE test.  i'd prefer individual tests for different use cases with some different parameters situation to try to capture edge cases.  Given the change in logic, I think this is showing not all scenarios are covered correctly by the tests.  AKA this change should have caused a test failure and didn't.  

InvalidRequestException comes from kork-exceptions which is included via kork-core.  SO no need for kork-web.  This provides actual feedback/better response handling through various paths as a rule and why we use it, or sometimes the error messages get lost across services.

Reply:
Created separate test cases for each valid and invalid cases